### PR TITLE
Support GpuHashJoin on Structs

### PR DESCRIPTION
**Pull request by sperlingxx**
Signed-off-by: sperlingxx <lovedreamf@gmail.com>

Current PR is to partially support GPU equal join (GpuHashJoin) on structure data (#2126). There are some constraints in current PR:
1. Due to some unresolved problems, we only support four join types: inner, left outer, right outer, cross, left anti and left semi. ~~Left anti and left semi join on structs have yet supported in libcudf ([#7912](https://github.com/rapidsai/cudf/issues/7912))~~. Full outer is not supported because of [#7934](https://github.com/rapidsai/cudf/issues/7934) and [#7947](https://github.com/rapidsai/cudf/issues/7947).
2. We can not support structs with float types, because Literal of struct type is not supported under GPU. #2140 
3. Sort merge join on nested structs are not supported, because `SortExec` is not supported with nested structs.
4. Current PR doesn't work with nullable structs until the cuDF fix PR [#7963](https://github.com/rapidsai/cudf/pull/7963) getting merged.

### CODE REVIEW

File: integration_tests/src/main/python/join_test.py
Comment (revans2) [inline integration_tests/src/main/python/join_test.py:59]: Why do we need to turn this on?

File: integration_tests/src/main/python/join_test.py
Comment (revans2) [inline integration_tests/src/main/python/join_test.py:122]: Instead of deleting these tests totally, can we update them so that the types of join we don't support are tested for fallback to the CPU?

File: integration_tests/src/main/python/join_test.py
Comment (revans2) [inline integration_tests/src/main/python/join_test.py:412]: Why not Right?

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuHashJoin.scala
Comment (revans2) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuHashJoin.scala:389]: nit: An explanation of why we need this might be helpful here. Right now we know that joins have issues with struct children, but it would be good to document what the issue is and if there are any follow on issues to address it.

File: integration_tests/src/main/python/join_test.py
Comment (sperlingxx) [reply integration_tests/src/main/python/join_test.py:59]: We turn on this because `GpuSortMeta` requires stable sort if there exists `StructType` in sortOrder. [link](https://github.com/NVIDIA/spark-rapids/blob/HEAD/sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOverrides.scala#L2882)

File: integration_tests/src/main/python/join_test.py
Comment (sperlingxx) [reply integration_tests/src/main/python/join_test.py:122]: Yes, I added them back with remaining unsupported join types.

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuHashJoin.scala
Comment (sperlingxx) [reply sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuHashJoin.scala:399]: Comments appended.

File: integration_tests/src/main/python/join_test.py
Comment (revans2) [reply integration_tests/src/main/python/join_test.py:59]: Then please add a comment here with a link to the issue that will fix it, so we can turn it off later. I just want to be sure that it is clear why a config is set, and that the out of core sort was not causing issues.

**Comment by revans2** [comment]
Looks really good just the one nit and then we can merge this in once the final fix is in place on the CUDF side.

File: integration_tests/src/main/python/join_test.py
Comment (sperlingxx) [reply integration_tests/src/main/python/join_test.py:59]: I have filed an issue for that: https://github.com/NVIDIA/spark-rapids/issues/2257

**Comment by sperlingxx** [comment]
build

**Comment by sperlingxx** [comment]
build

File: integration_tests/src/main/python/join_test.py
Comment (sperlingxx) [reply integration_tests/src/main/python/join_test.py]: We got rid of this configuration after #2092 merged.

**Comment by sperlingxx** [comment]
build

**Comment by sperlingxx** [comment]
build

**Comment by sperlingxx** [comment]
build

**Comment by sperlingxx** [comment]
> Looks really good just the one nit and then we can merge this in once the final fix is in place on the CUDF side.

The fix on the CUDF side is done. And all checks of CI pipeline are happy.

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuHashJoin.scala
Comment (revans2) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuHashJoin.scala:44]: nit:

I personally find it much more readable to do a def instead of a val for functions like this.

```
def unSupportNonEqualCondition:
  ...

def unSupportStructKeys:
  ...
```

**Review by revans2 — APPROVED** [review]
I had one really minor style nit, I don't expect you to fix it. It is small enough I am just going to ignore it.

**Comment by revans2** [comment]
You had to do the build 4 times to get it to pass. Why exactly did https://blossom.nvidia.com/sw-gpu-spark-jenkins/job/rapids_premerge-github/1454/ fail (with what appears to be a join issue) and what changed to make it pass? I don't see anything that changed from CI. Which makes me really nervous.

**Review by revans2 — CHANGES_REQUESTED** [review]
Removing my approval until I understand what happened with CI

File: shims/spark300/src/main/scala/com/nvidia/spark/rapids/shims/spark300/GpuBroadcastHashJoinExec.scala
Comment (revans2) [inline shims/spark300/src/main/scala/com/nvidia/spark/rapids/shims/spark300/GpuBroadcastHashJoinExec.scala:146]: Sorry it looks like we missed the databricks shim for this change too.

Comment (sperlingxx) [reply shims/spark300/src/main/scala/com/nvidia/spark/rapids/shims/spark300/GpuBroadcastHashJoinExec.scala:146]: Oh, my bad! Appended.

**Comment by sperlingxx** [comment]
> You had to do the build 4 times to get it to pass. Why exactly did https://blossom.nvidia.com/sw-gpu-spark-jenkins/job/rapids_premerge-github/1454/ fail (with what appears to be a join issue) and what changed to make it pass? I don't see anything that changed from CI. Which makes me really nervous.

The CI failed several times due to this cuDF [bug](https://github.com/rapidsai/cudf/pull/8075). I wasn't aware of this bug until I found the CI still failed after multiple attempts. (I just thought it is because the cuDF Jar hasn't been updated to latest master.  Therefore, I tried to rerun CI hours later.) Then, I looked into the error log and found the problem.

**Comment by sperlingxx** [comment]
build

**Review by revans2 — APPROVED** [review]

**Pull request merged**
